File: exercise/scala/day22/src/test/scala/EIDProperties.scala
import Generators.validEID
import Mutators.mutantGenerator
import eid.EID
import org.scalacheck.Prop.{classify, forAll}
import org.scalacheck.{Gen, Properties}
import org.scalatest.EitherValues

case class Mutator(name: String, func: EID => Gen[String]) {
  def mutate(eid: EID): String = func(eid).sample.get
}

object EIDProperties extends Properties("EID") with EitherValues {

  property("round tripping") = forAll(validEID) { eid =>
    EID.parse(eid.toString).value == eid
  }

  property("Invalid EID can never be parsed") = forAll(validEID, mutantGenerator) { (eid, mutator) =>
    classify(true, mutator.name) {
      EID.parse(mutator.mutate(eid)).isLeft
    }
  }
}

object Mutators {
  val aMutator: Mutator = Mutator("a mutator", eid => Gen.const("Implement this first mutator"))

  val mutantGenerator: Gen[Mutator] = Gen.oneOf(
    aMutator,
    aMutator
  )
}

File: exercise/kotlin/day22/src/test/kotlin/EIDProperties.kt
package eid.tests

import eid.EID
import eid.SerialNumber
import eid.Sex
import eid.Year
import io.kotest.core.spec.style.StringSpec
import io.kotest.property.Arb
import io.kotest.property.arbitrary.arbitrary
import io.kotest.property.arbitrary.enum
import io.kotest.property.arbitrary.int
import io.kotest.property.arbitrary.of
import io.kotest.property.forAll

val eidGenerator = arbitrary { _ ->
    EID(
        sex = Arb.enum<Sex>().bind(),
        year = Year(Arb.int(0, 99).bind()),
        serialNumber = SerialNumber(Arb.int(1, 999).bind())
    )
}

class EIDProperties : StringSpec({
    "EID round tripping" {
        forAll(eidGenerator) { validEID ->
            EID.parse(validEID.toString()).getOrNull() == validEID
        }
    }

    "Invalid EID can never be parsed" {
        forAll(eidGenerator, mutantGenerator) { validEID, mutator ->
            collect(mutator.name)
            EID.parse(mutator.mutate(validEID)).isFailure
        }
    }
})

private data class Mutator(val name: String, val func: (EID) -> Arb<String>) {
    fun mutate(eid: EID): String = func(eid).samples().first().value
}

private val aMutator: Mutator = Mutator("a mutator") { eid ->
    Arb.of("Implement this first mutator")
}

private val mutantGenerator: Arb<Mutator> = Arb.of(
    aMutator
)